// zcatalyst-sdk-node/lib/email/email.js
'use strict';
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Email = void 0;
const validator_1 = require("../utils/validator");
const emailValidator_1 = require("./emailValidator");
const api_request_1 = require("../utils/api-request");
const error_1 = require("../utils/error");
const constants_1 = __importDefault(require("../utils/constants"));
const { REQ_METHOD, COMPONENT } = constants_1.default;
class Email {
    constructor(app) {
        (0, validator_1.isValidApp)(app, true);
        this.requester = new api_request_1.AuthorizedHttpClient(app, this);
    }
    getComponentName() {
        return COMPONENT.email;
    }
    sendMail(mailObj) {
        return __awaiter(this, void 0, void 0, function* () {
            yield (0, validator_1.wrapValidatorsWithPromise)(() => {
                (0, validator_1.isNonEmptyObject)(mailObj, 'email_object', true);
            }, error_1.CatalystEmailError);
            const validMailObj = yield (0, emailValidator_1.validate)(mailObj);
            const request = {
                method: REQ_METHOD.post,
                path: `/email/send`,
                data: validMailObj,
                type: 'json',
                catalyst: true,
                track: true
            };
            const resp = yield this.requester.send(request);
            return resp.data.data;
        });
    }
}
exports.Email = Email;
